GHC 8.0.2-9448e62: Couldn't match type ‘m’ with ‘StateT s m’

purescript 0.10.2 build failure with GHC pre-8.0.2-9448e62 (https://github.com/ghc/ghc/commit/9448e62740ca03aeb915bf0ecf8b16e54a52798a).

```
Building purescript-0.10.2...
Preprocessing library purescript-0.10.2...
[  1 of 139] Compiling System.IO.UTF8   ( src/System/IO/UTF8.hs, dist/dist-sandbox-4bbb699a/build/System/IO/UTF8.o )
[  2 of 139] Compiling Paths_purescript ( dist/dist-sandbox-4bbb699a/build/autogen/Paths_purescript.hs, dist/dist-sandbox-4bbb699a/build/Paths_purescript.o )
[  3 of 139] Compiling Language.PureScript.Traversals ( src/Language/PureScript/Traversals.hs, dist/dist-sandbox-4bbb699a/build/Language/PureScript/Traversals.o )
[  4 of 139] Compiling Language.PureScript.Publish.Utils ( src/Language/PureScript/Publish/Utils.hs, dist/dist-sandbox-4bbb699a/build/Language/PureScript/Publish/Utils.o )
[  5 of 139] Compiling Language.PureScript.Publish.BoxesHelpers ( src/Language/PureScript/Publish/BoxesHelpers.hs, dist/dist-sandbox-4bbb699a/build/Language/PureScript/Publish/BoxesHelpers.o )
[  6 of 139] Compiling Language.PureScript.Parser.State ( src/Language/PureScript/Parser/State.hs, dist/dist-sandbox-4bbb699a/build/Language/PureScript/Parser/State.o )
[  7 of 139] Compiling Language.PureScript.Options ( src/Language/PureScript/Options.hs, dist/dist-sandbox-4bbb699a/build/Language/PureScript/Options.o )
[  8 of 139] Compiling Language.PureScript.Kinds ( src/Language/PureScript/Kinds.hs, dist/dist-sandbox-4bbb699a/build/Language/PureScript/Kinds.o )
[  9 of 139] Compiling Language.PureScript.Interactive.IO ( src/Language/PureScript/Interactive/IO.hs, dist/dist-sandbox-4bbb699a/build/Language/PureScript/Interactive/IO.o )
[ 10 of 139] Compiling Language.PureScript.Docs.Utils.MonoidExtras ( src/Language/PureScript/Docs/Utils/MonoidExtras.hs, dist/dist-sandbox-4bbb699a/build/Language/PureScript/Docs/Utils/MonoidExtras.o )
[ 11 of 139] Compiling Language.PureScript.Crash ( src/Language/PureScript/Crash.hs, dist/dist-sandbox-4bbb699a/build/Language/PureScript/Crash.o )
[ 12 of 139] Compiling Language.PureScript.Comments ( src/Language/PureScript/Comments.hs, dist/dist-sandbox-4bbb699a/build/Language/PureScript/Comments.o )
[ 13 of 139] Compiling Language.PureScript.Parser.Lexer ( src/Language/PureScript/Parser/Lexer.hs, dist/dist-sandbox-4bbb699a/build/Language/PureScript/Parser/Lexer.o )
[ 14 of 139] Compiling Language.PureScript.Bundle ( src/Language/PureScript/Bundle.hs, dist/dist-sandbox-4bbb699a/build/Language/PureScript/Bundle.o )

src/Language/PureScript/Bundle.hs:535:5: warning: [-Wincomplete-patterns]
    Pattern match(es) are non-exhaustive
    In an equation for ‘squash’: Patterns not matched: JSAnnotSpace

src/Language/PureScript/Bundle.hs:590:7: warning: [-Wincomplete-patterns]
    Pattern match(es) are non-exhaustive
    In an equation for ‘appendAnn’:
        Patterns not matched: _ JSAnnotSpace
[ 15 of 139] Compiling Language.PureScript.AST.SourcePos ( src/Language/PureScript/AST/SourcePos.hs, dist/dist-sandbox-4bbb699a/build/Language/PureScript/AST/SourcePos.o )
[ 16 of 139] Compiling Language.PureScript.AST.Operators ( src/Language/PureScript/AST/Operators.hs, dist/dist-sandbox-4bbb699a/build/Language/PureScript/AST/Operators.o )
[ 17 of 139] Compiling Language.PureScript.AST.Literals ( src/Language/PureScript/AST/Literals.hs, dist/dist-sandbox-4bbb699a/build/Language/PureScript/AST/Literals.o )
[ 18 of 139] Compiling Control.Monad.Supply ( src/Control/Monad/Supply.hs, dist/dist-sandbox-4bbb699a/build/Control/Monad/Supply.o )
[ 19 of 139] Compiling Control.Monad.Supply.Class ( src/Control/Monad/Supply/Class.hs, dist/dist-sandbox-4bbb699a/build/Control/Monad/Supply/Class.o )

src/Control/Monad/Supply/Class.hs:30:10: error:
    • Couldn't match type ‘m’ with ‘StateT s m’
      ‘m’ is a rigid type variable bound by
        the instance declaration at src/Control/Monad/Supply/Class.hs:30:10
      Expected type: StateT s m Integer
        Actual type: StateT s (StateT s m) Integer
    • In the expression:
        Control.Monad.Supply.Class.$dmfresh @StateT s m
      In an equation for ‘fresh’:
          fresh = Control.Monad.Supply.Class.$dmfresh @StateT s m
      In the instance declaration for ‘MonadSupply (StateT s m)’
    • Relevant bindings include
        fresh :: StateT s m Integer
          (bound at src/Control/Monad/Supply/Class.hs:30:10)

src/Control/Monad/Supply/Class.hs:30:10: error:
    • Couldn't match type ‘m’ with ‘StateT s m’
      ‘m’ is a rigid type variable bound by
        the instance declaration at src/Control/Monad/Supply/Class.hs:30:10
      Expected type: StateT s m Integer
        Actual type: StateT s (StateT s m) Integer
    • In the expression: Control.Monad.Supply.Class.$dmpeek @StateT s m
      In an equation for ‘peek’:
          peek = Control.Monad.Supply.Class.$dmpeek @StateT s m
      In the instance declaration for ‘MonadSupply (StateT s m)’
    • Relevant bindings include
        peek :: StateT s m Integer
          (bound at src/Control/Monad/Supply/Class.hs:30:10)

src/Control/Monad/Supply/Class.hs:31:10: error:
    • Couldn't match type ‘m’ with ‘WriterT w m’
      ‘m’ is a rigid type variable bound by
        the instance declaration at src/Control/Monad/Supply/Class.hs:31:10
      Expected type: WriterT w m Integer
        Actual type: WriterT w (WriterT w m) Integer
    • In the expression:
        Control.Monad.Supply.Class.$dmfresh @WriterT w m
      In an equation for ‘fresh’:
          fresh = Control.Monad.Supply.Class.$dmfresh @WriterT w m
      In the instance declaration for ‘MonadSupply (WriterT w m)’
    • Relevant bindings include
        fresh :: WriterT w m Integer
          (bound at src/Control/Monad/Supply/Class.hs:31:10)

src/Control/Monad/Supply/Class.hs:31:10: error:
    • Couldn't match type ‘m’ with ‘WriterT w m’
      ‘m’ is a rigid type variable bound by
        the instance declaration at src/Control/Monad/Supply/Class.hs:31:10
      Expected type: WriterT w m Integer
        Actual type: WriterT w (WriterT w m) Integer
    • In the expression:
        Control.Monad.Supply.Class.$dmpeek @WriterT w m
      In an equation for ‘peek’:
          peek = Control.Monad.Supply.Class.$dmpeek @WriterT w m
      In the instance declaration for ‘MonadSupply (WriterT w m)’
    • Relevant bindings include
        peek :: WriterT w m Integer
          (bound at src/Control/Monad/Supply/Class.hs:31:10)
cabal: Leaving directory '.'
cabal: Error: some packages failed to install:
purescript-0.10.2 failed during the building phase. The exception was:
ExitFailure 1

```

Full build log: https://bot.brew.sh/job/Homebrew%20Core/10983/version=any_sierra/consoleText

(link will eventually die, so duplicated here: https://gist.githubusercontent.com/ilovezfs/fb1c282b2a123e23d480882e74f74db4/raw/24f75f77fa0922d593ec58df7a6787314a9c54e8/gistfile1.txt)

--- a/CONTRIBUTORS.md
+++ b/CONTRIBUTORS.md
@@ -84,6 +84,7 @@ This file lists the contributors to the PureScript compiler project, and the ter
 - [@zudov](https://github.com/zudov) (Konstantin Zudov) My existing contributions and all future contributions until further notice are Copyright Konstantin Zudov, and are licensed to the owners and users of the PureScript compiler project under the terms of the [MIT license](http://opensource.org/licenses/MIT).
 - [@brandonhamilton](https://github.com/brandonhamilton) (Brandon Hamilton) My existing contributions and all future contributions until further notice are Copyright Brandon Hamilton, and are licensed to the owners and users of the PureScript compiler project under the terms of the [MIT license](http://opensource.org/licenses/MIT).
 - [@bbqbaron](https://github.com/bbqbaron) (Eric Loren) My existing contributions and all future contributions until further notice are Copyright Eric Loren, and are licensed to the owners and users of the PureScript compiler project under the terms of the [MIT license](http://opensource.org/licenses/MIT).
+- [@RyanGlScott](https://github.com/RyanGlScott) (Ryan Scott) My existing contributions and all future contributions until further notice are Copyright Ryan Scott, and are licensed to the owners and users of the PureScript compiler project under the terms of the [MIT license](http://opensource.org/licenses/MIT).
 
 ### Companies
 

--- a/src/Control/Monad/Supply/Class.hs
+++ b/src/Control/Monad/Supply/Class.hs
@@ -3,6 +3,7 @@
 --
 
 {-# LANGUAGE DefaultSignatures #-}
+{-# LANGUAGE TypeFamilies #-}
 
 module Control.Monad.Supply.Class where
 
@@ -15,9 +16,9 @@ import Control.Monad.Writer
 class Monad m => MonadSupply m where
   fresh :: m Integer
   peek :: m Integer
-  default fresh :: MonadTrans t => t m Integer
+  default fresh :: (MonadTrans t, MonadSupply n, m ~ t n) => m Integer
   fresh = lift fresh
-  default peek :: MonadTrans t => t m Integer
+  default peek :: (MonadTrans t, MonadSupply n, m ~ t n) => m Integer
   peek = lift peek
 
 instance Monad m => MonadSupply (SupplyT m) where

